### PR TITLE
chore: add test coverage

### PR DESCRIPTION
**Pull request by bcoe**
Adds test coverage using `nyc` and `babel-plugin-istanbul`, see #244

screen shot 2016-11-15 at 12 35 24 am

## a few notes about this pull request

* I've switched to using `babel-core/register`, which is easier to wire up with `babel-plugin-istanbul`.
* I've moved babel settings into `.babelrc`, and in turn load these in `gulpfile.js`; both `babel-register`, and the build process need these same settings, and it seemed worthwhile to consolidate them.
* I've introduced a few more dependencies:
  * `coveralls`: used to pipe lcov report to coveralls.io.
  * `read-pkg-up`: used to load the parent package.json (this is necessary because in test files are under `lib/`, but after building they are located in `dist/lib`.
  * `babel-plugin-istanbul`: istanbul plugin for instrumenting test coverage (only run when env is `test`).
  * `nyc`: command line tool for instrumenting tests, outputting reports, etc.
  * `cross-env`: pass environment to a script in a way that works on Windows.
* I added a couple more tests, mainly to ensure that coverage was being computed accurately.

## enabling coverage badge

To get the coverage bade working, someone with admin access to this repo will need to enable this project on https://coveralls.io/

### CODE REVIEW

**Comment by bcoe** [comment]
Just a quick update, I've added a few scripts in this pull request:

* `npm test`: runs tests through gulp, pretty much the same as the old behavior.
* `npm run ava`: run ava in isolation, this is good thing to have as we build out tests, so that we don't need to run the full suite each time; I was for instance doing this while testing:
  
screen shot 2016-11-15 at 10 34 05 am

* `npm run html-report`: output a pretty HTML report, which makes it easier to hunt down lines of code missing coverage:

screen shot 2016-11-15 at 10 35 13 am

* `npm run coveralls`: pipe coverage output, from the last test run, to coveralls -- this is called in `.travis.yml`.

**Comment by nkzawa** [comment]
Thanks, LGTM ! 😄

**Review by rauchg — CHANGES_REQUESTED** [review]

**Comment by bcoe** [comment]
@rauchg I agree that this felt a bit weird; part of me wants to set NODE_ENV=test in nyc, but this feels like it's not flexible enough -- I've instead pulled in the neat little plugin `gulp-process-env`, which feels pretty elegant.

File: gulpfile.js [hunk 1]
```diff
@@ -153,9 +133,13 @@ gulp.task('build-client', ['compile-lib', 'compile-client'], () => {
   .pipe(notify('Built release client'))
 })
 
-gulp.task('test', ['compile', 'copy', 'compile-test', 'copy-test-fixtures'], () => {
-  return gulp.src('dist/test/*.js')
-  .pipe(ava())
+gulp.task('test', () => {
+  process.env.NODE_ENV = 'test'
```
Comment (arunoda) [inline gulpfile.js]: I think this is fine.
But we could do this too.

```
process.env.NODE_ENV = process.env.NODE_ENV || 'test'
```

**Comment by arunoda** [comment]
Guys, I wonder the state of this PR with the potential decision(https://github.com/zeit/next.js/issues/278) to move to Jest.

**Comment by rauchg** [comment]
I want to merge this sooner rather than later. Shouldn't be hard to substitute the runner, since we're basically making assertions of snapshots already

**Comment by bcoe** [comment]
@rauchg @arunoda rebased again, in case you do decide to land this; I'd be happy to help with the Jest switch over when the time comes -- but I would like to dig into whether there's a way to make Jest's use of `babel-plugin-istanbul` play nice with `nyc` ... it would be nice to have access to the lovely advanced reporting features.

**Comment by nkzawa** [comment]
LGTM

**Comment by bcoe** [comment]
@nkzawa @rauchg follow up, make sure you registry the project here:

https://coveralls.io/r/zeit/next.js?branch=master

and kick off a build on `master`, otherwise the bade won't populate.

**Comment by nkzawa** [comment]
@bcoe registered, thanks!

**Pull request merged**